src/sage/algebras/quatalg/quaternion_algebra.py

```diff
@@ -2256,6 +2259,291 @@ def isomorphism_to(self, other, *, conjugator=False):
+
+        return return_map
+
+    def _random_left_ideal_2q(self, N):
+        r"""
+        Return random left ideals of norm `N` when :meth:`modp_splitting_map`
+        is not implemented, i.e. when `N \mid 2i^2j^2`.
+
+        This method is slow when `N` is large.
+
+        EXAMPLES::
+
+            sage: B = QuaternionAlgebra(-7, -13)
+            sage: O = B.maximal_order()
+            sage: O._random_left_ideal_2q(2).norm() == 2
+            True
+            sage: O._random_left_ideal_2q(7).norm() == 7
+            True
+            sage: O._random_left_ideal_2q(13).norm() == 13
+            True
+        """
+        alpha = self.random_element()
+        while gcd(alpha.reduced_norm(), N*N) != N:
+            alpha = self.random_element()
+        return self*alpha + self*N
+
+    def random_left_ideal_of_norm(self, N, primitive=True):
+        r"""
+        Return a random left ideal of ``self`` of norm `N`.
+
+        INPUT:
+
+        - ``N`` -- (integer) the norm of the ideal returned
+        - ``primitive`` -- (bool, default: ``True``) if set, the output ideal
+          is guaranteed to be primitive
+
+        OUTPUT: a left ideal of self of chosen norm `N`
+
+        .. WARNING::
+
+            The algorithm requires factoring `N`, so for large input the algorithm
+            will be slow.
+
+        EXAMPLES::
+
+            sage: B = QuaternionAlgebra(next_prime(2**50))
+            sage: O = B.maximal_order()
+            sage: N = 2**10 * 3**10 * 5**10
+            sage: assert O.random_left_ideal_of_norm(N).norm() == N
+
+        Verify that the method works when N is not coprime to `i^2` or `j^2`::
+
+            sage: B = QuaternionAlgebra(-7, -13)
+            sage: O = B.maximal_order()
+            sage: N = 3**5 * 7**5
+            sage: O.random_left_ideal_of_norm(N).norm() == N
+            True
+
+        Verify that the method works when N is not coprime to `i^2` or `j^2`::
+
+            sage: B = QuaternionAlgebra(-7, -13)
+            sage: O = B.maximal_order()
+            sage: N = 3**5 * 7**5
+            sage: O.random_left_ideal_of_norm(N).norm() == N
+            True
+
+        TESTS::
+
+            sage: for _ in range(10):
+            ....:     p, q = [random_prime(100, lbound=10) for _ in range(2)]
+            ....:     ep, eq = [choice([-1, 1]) for _ in range(2)]
+            ....:     B = QuaternionAlgebra(ep * p, eq * q)
+            ....:     if not B.is_division_algebra():
+            ....:         continue
+            ....:     O = B.maximal_order()
+            ....:     N = random_prime(100)
+            ....:     try:
+            ....:         assert O.random_left_ideal_of_norm(N).norm() == N
+            ....:     except (NotImplementedError, ValueError, TypeError, ZeroDivisionError):
+            ....:         pass
+        """
+        if self.base_ring() != ZZ:
+            raise NotImplementedError("only implemented for quaternion algebras over ZZ")
```

Comment:
Since ZZ is the base ring of the order, not of the quaternion algebra:
```suggestion
        if self.base_ring() != ZZ:
            raise NotImplementedError("only implemented for rational quaternion algebras")
```

Or, to be more in line with the code some lines below (although I prefer the error message above):
```suggestion
        if self.base_ring() != ZZ:
            raise NotImplementedError("only implemented for quaternion algebras over QQ")
```